packages/transport-webrtc/test/certificate-persistent.spec.ts

```diff
@@ -0,0 +1,176 @@
+ * 
+ * @author Crosstons
+ * @date 2025-03-11 10:09:19
+ */
+
+import { expect } from 'aegir/chai'
+import { MemoryDatastore } from 'datastore-core/memory'
+import type { Keychain } from '@libp2p/keychain'
+import { createLibp2p } from 'libp2p'
+import { webRTCDirect } from '../src/index.js'
+import { CERTIFICATE_KEY_PREFIX } from '../src/private-to-public/utils/certificate-store.js'
+import { mplex } from '@libp2p/mplex'
+import { noise } from '@chainsafe/libp2p-noise'
+import { logger } from '@libp2p/logger'
+import { Key } from 'interface-datastore/key'
+import * as sinon from 'sinon'
+
+// Create a properly namespaced logger for these tests
+const testLogger = logger('libp2p:webrtc:test:certificate-persistence')
+
+describe('WebRTC Certificate Persistence', () => {
+  let sandbox: sinon.SinonSandbox
+  
+  before(() => {
+    // Create sinon sandbox for stubbing functions
+    sandbox = sinon.createSandbox()
+  })
+  
+  afterEach(() => {
+    // Restore all stubbed functions
+    sandbox.restore()
+  })
+  
+  describe('certificate integration', function () {
+    // These tests might take longer to run
+    this.timeout(20000)
+
+    // Helper function to create a node with keychain
+    async function createNode(persistentDatastore: MemoryDatastore | null = null, options = {}) {
+      const nodeDatastore = persistentDatastore || new MemoryDatastore()
+      
+      return await createLibp2p({
+        addresses: {
+          listen: ['/ip4/127.0.0.1/udp/0/webrtc-direct']
+        },
+        datastore: nodeDatastore,
+        connectionGater: {
+          denyDialMultiaddr: async () => false
+        },
+        streamMuxers: [mplex()],
+        connectionEncrypters: [noise()],
+        transports: [
+          webRTCDirect({
+            ...options,
+            dataChannel: {
+              maxMessageSize: 1 << 16
+            }
+          })
+        ],
+        services: {
+          keychain: async (components) => {
+            const keychainModule = await import('@libp2p/keychain')
+            return keychainModule.keychain({
+              pass: 'very-secure-password-for-testing',
+              dek: {
+                keyLength: 512 / 8,
+                iterationCount: 1000,
+                salt: 'at-least-16-characters-long-for-testing',
+                hash: 'sha2-512'
+              }
+            })(components)
+          }
+        }
+      })
+    }
+    
+    it('should use certificate hash in multiaddr', async () => {
+      // Create a node with certificate persistence enabled
+      const node = await createNode(null, {
+        certificates: [], // Empty array to not use any pre-defined certificates
+        useLibjuice: false, // Use WebRTC's built-in STUN/TURN
+        rtcConfiguration: {
+          iceTransportPolicy: 'all',
+        }
+      })
+      
+      try {
+        await node.start()
+
+        // Wait for addresses to be available
+        await new Promise(resolve => setTimeout(resolve, 1000))
```

Comment:
Please don't add arbitrary delays, they make the tests take a long time to run and can be flakey.

Instead, use something like `p-wait-for` and test for the thing you are waiting for.